utils/utf8.cc

```diff
@@ -584,6 +586,30 @@ std::optional<size_t> validate_with_error_position(const uint8_t *data, size_t l
     return validate_naive(data, len);
 }
 
+// From https://stackoverflow.com/a/33799784/12786387
+sstring escape_json(const sstring &s) {
+    std::ostringstream o;
+    for (auto c = s.cbegin(); c != s.cend(); c++) {
```

Comment:
Please check if we can reuse something from rapidjson.

Reply:
btw, it's probably better to check first if escaping needs to be done, rather than doing it unconditionally.

Reply:
See my other comment about this code should move into Seastar (I explained where), but there, please add a unit test for it, to make sure the code actually does what you think it does for various examples (quotes, newlines, escape sequences, and so on).

Reply:
By the way, I compared your code to the JSON RFC (https://github.com/scylladb/seastar/issues/460) and it seems to me you're escaping exactly the right things, so good.

Reply:
@nyh in my original bug you said that in Alternator you are parsing the strings using rapid json, why not do the same here?

Reply:
Alternator keeps JSON objects as RapidJson objects (`rjson::value` is our wrapper from utikls/rjson.hh) but converts it to a string before passing it to Seastar.

But here we're talking about code "written" by your API-to-code converter, and this code keeps the JSON in memory using some home-brewed "json_element" code, and it's converted into a string using buggy code in Seastar.

If we keep the API-to-code converter code as, is, we need to fix the buggy conversion code in Seastar. I explained in https://github.com/scylladb/seastar/issues/460 exactly where (I think) the fix needs to go in, and the fix itself would be the same escaping code that @jul-stas added here.

Reply:
@nyh my question is about scylladb/seastar#460 I agree the code in seastart needs to be change, that's why I opened it in the first place, I'm asking if we can't use rapid-json in seastar to do the escaping.

Reply:
As I said, Seastar doesn't use rapidjson and instead uses its own home-brewed "json_element" stuff. We can indeed take out all that `json_element` and `to_json()` stuff and replace it with rapidjson.
I suggested something even more radical: I suggested we should take out all the `json_element` and `to_json()` stuff out of Seastar, and not replace it with anything else - Seastar will simply not know anything about JSON, and it's the application's job to convert whatever it has into a string which is what HTTP knows about. The application can use whatever JSON library it wants, including rapidjson or anything else.

But I think my short-term proposal is to fix the `formatter::to_json(char*)` function in Seastar. The longer-term reorganization can come later, and will also cause backward-compatibility problems so we can't remove it without some sort of deprecation period.